kube/src/runtime/controller.rs

```diff
@@ -0,0 +1,178 @@
+            WatchEvent::Error(e) => Err(Error::Api(e)),
+        }
+    }
+}
+
+/// An Ok return value from a reconcile fn
+///
+/// Designed so the Controller can decide  whether to requeue the event
+/// Error cases are not encapsulated in this struct (they are handled by Result)
+#[derive(Debug)]
+pub enum ReconcileStatus {
+    /// Successful reconcile
+    Complete,
+    ///  Partial success, reque after some time
+    RequeAfter(Duration),
+}
+
+/// A controller for a kubernetes object K
+#[derive(Clone)]
+pub struct Controller<K, F>
+where
+    K: Clone + DeserializeOwned + Meta,
+    F: Fn(ReconcileEvent) -> Result<ReconcileStatus>,
+{
+    client: APIClient,
+    resource: Resource,
+    informers: Vec<Informer<K>>,
+    reconciler: Box<F>,
+}
+
+// TODO: is 'static limiting here?
+impl<K: 'static, F: 'static> Controller<K, F>
+where
+    K: Clone + DeserializeOwned + Meta + Send + Sync,
+    F: Fn(ReconcileEvent) -> Result<ReconcileStatus> + Send + Sync,
+{
+    /// Create a controller with a kube client on a kube resource
+    pub fn new(client: APIClient, r: Resource, recfn: F) -> Self {
+        Controller {
+            client: client,
+            resource: r,
+            informers: vec![],
+            reconciler: Box::new(recfn),
+        }
+    }
+
+    /// Create internal informers for an associated kube resource
+    ///
+    /// TODO: this needs to only find resources with a property matching root resource
+    pub fn owns(mut self, r: Resource, lp: ListParams) -> Self {
+        self.informers.push(Informer::new(self.client.clone(), lp, r));
+        self
+    }
+
+    /// Initialize
+    pub fn init(self) {
+        info!("Starting Controller for {:?}", self.resource);
+        let (tx, mut rx) = mpsc::unbounded_channel();
+
+        // 1. poll informers in parallel and push results to queue
+        for inf in self.informers.clone() {
+            // TODO: ownership move?
+            //let queue = self.queue.clone();
+            let txi = tx.clone();
+            tokio::spawn(async move {
+                let mut poll_i = inf.poll().await.expect("could talk to api server").boxed();
+                while let Some(ev) = poll_i.next().await {
+                    match ev {
+                        Ok(wi) => {
+                            let ri = ReconcileEvent::try_from(wi);
+                            //(*queue.lock().await).push_back(ri);
+                            txi.send(ri).expect("channel can receive");
+                        }
+                        _ => unimplemented!(),
+                        //Err(e) => tx.unbounded_send(Err(e)),
+                    }
+                }
+            });
+        }
+        // TODO: init main informer
+        // TODO: queue up events
+        // TODO: debounce events
+
+
+        // Prepare a sender so we can stack things back on the channel
+        let txa = tx.clone();
+        // Event loop that triggers the reconcile fn
+        tokio::spawn(async move {
```

Comment:
Makes sense while prototyping.. but we should try to avoid `tokio::spawn` before releasing, because it means that `kube-rs` has to take complete control of the error handling. Should be possible to replace most uses with some variant of [`try_join!`](https://docs.rs/futures/0.3.4/futures/macro.try_join.html).